Fix world memory leak

TheDGOfficial opened the pull request:
![image](https://github.com/user-attachments/assets/e0785983-2376-4a0b-9ba3-bcfc2c02a511)

Fixes a world memory leak, which caused insuffurable lag due to many Old Garbage Collections after long gameplay sessions involving swapping lots of worlds, due to CommandSender (mostly EntityPlayerSP, ourselves) keeping a reference to the world they were in, even after changing to a new world.

A new EntityPlayerSP instance is created and the old one should be discarded but NEU kept a strong reference to it, and the player instance has a reference to world, and the world of course has references to everything inside it; chunks, entities, tile entities, etc. - which caused all the previous worlds you have been in to not be cleared from memory. The leak is not noticeable in effect at start before you swap worlds, and you probably wouldn't notice after a few swaps either. But if you play for a long session and swap lot of worlds, such as Dungeon instances, your memory will continue filling up and require a restart to get rid of 3 second plus Old Gen GC memory cleanup done by the JVM every few seconds (which is a Stop-The-World process, so the game will freeze during this memory cleanup performed by JVM, in simple terms.).

If I'm going to be honest, this fix is not the best, but it fixes the leak. Possible issues with the fix:
 - WeakReference.get can return null, old code had a strong reference so it would never be null. Not sure if this could create an issue, but all NEU commands seem to work correctly still per my testing. Likely because .get is only ever called with the current player (due to it being impossible to send commands from a world you just left).
 -  WeakReference object itself is still leaked even if the underlying CommandSender (and the World it's in) isn't. While this consumes much less memory than before (a whole world with all its entities, chunks, etc. being leaked vs a small wrapper object being leaked) it's still not a clean solution. A clean solution would be to use a Weak collection type that automatically discards cleared references, such as WeakHashMap from Java standard library. I am not sure we should be replacing the LRUCache though, as I didn't write this code I don't know if changing it to a WeakHashMap would introduce any further bugs or regressions.
 - WeakReference.hashCode is likely different, we create a new WeakReference each time a CommandSender is put to the LRUCache, it probably doesn't return the underlying CommandSender's hashCode so this might break the purpose of the cache entirely. However LRUCache seems to be a NEU internal class and not a standard Java collection type so I don't know if it depends on .hashCode for its operations.
 - Not sure why the NEUBrigadierHook.kt file is in src/main/java instead of src/main/kotlin, but I didn't want to change that in an unrelated PR.
 - I have not tested other CommandSender implementations but I assume that's a very rare use case anyways (such as CommandBlockCommandSender in singleplayer)
 
Tested different NEU commands including tab completion and everything seems to work after the fix and no leak is present, with no Old GC occurring.

Input needed regarding Point no 2 and 3 about a possible change to WeakHashMap, which would both fix WeakReference wrapper objects still being leaked and the .hashCode issue. Moving the file from src/main/java to src/main/kotlin is out of scope in this PR, point no. 1 likely doesnt cause any issues, and point no 5. likely does not apply since I doubt NEU supports running it's commands from a command block at all.

PR marked as draft till possible change to WeakHashMap or ensuring WeakReference does not cause any of the possible listed issues.

Slight footnote: I have said world, but assume it's a BungeeCord server change in Hypixel. I haven't tested this, but the EntityPlayerSP is likely 1 static object if you are playing in a regular server or singleplayer even if you change worlds.

TheDGOfficial committed:
Fix world memory leak

TheDGOfficial commented:
After checking out source code of LRUCache, this cache seems to be created with 1 entry limit. The leak might be in the lambda capturing the command sender in that case, or something else might be going on here.  I am not very sure. Input wanted. @nea89o Tagging since you are the author of the Brigadier and LRUCache files.

I will later revert the fix locally to reproduce the leak again and take a heapdump and analyze again what's going on since I deleted the original heapdump.

lineargraph commented:
LGTM. was aware of the leak of the world reference (since over a year ago), didnt realize the LRU cache i used had a cache history of more than one so i thought it was fine (since running any command in the new world would ditch the old world (just a bit later)).

TheDGOfficial committed:
Use WeakHashMap instead of WeakReference

- Fixes .hashCode being different due to a new WeakReference being created everytime an entry is put into cache, as WeakHashMap now handles that automatically for us.
- Fixes WeakReference wrapper object still being leaked even though the object it holds reference into is not. WeakHashMap automatically discards entries of discarded (Garbage Collected) references for us.

Also made a very slight change to fix an instance where NEU was saving a local variable redirect (assigned to value of node.redirect) and doing a null check on it but not using the local variable and instead using node.redirect. Changed it to use the null-checked local variable.

Kept the Collections.synchronizedMap used in LRUCache in the old code for thread safety (wraps the WeakHashMap with the .sychronizedMap), but I do not know if that's needed since MC is single threaded, but I'd suppose its best to not change behaviour of thread-safety.

NopoTheGamer reviewed (approved):
didnt test
